src/core/unit.c

```diff
@@ -2025,6 +2126,10 @@ int unit_start(Unit *u, ActivationDetails *details) {
         if (!u->activation_details) /* Older details object wins */
                 u->activation_details = activation_details_ref(details);
 
+        r = unit_increment_instance_count(u);
```

Comment:
this is the wrong place, as unit_start() is only called on explicit start requests of this unit. But units might change state for extenral reasons too, not just because systemd wants so.

hence, please put this in unit_notify() which is called whenever units change state.